### PR TITLE
44184 Handle dead letter (distributor failure) by removing the subscription

### PR DESCRIPTION
**Pull request by ganpat-tech**
PR contain new blob trigger on the existing webjob.
Code changes ,unit test case and automated test case.

### CODE REVIEW

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/appsettings.json [hunk 1]
```diff
@@ -47,14 +48,16 @@
...
         "D365ApiUri": "",
         "SucceededStatusCode": 832930001,
-        "FailedStatusCode": 832930002,        
+        "FailedStatusCode": 832930002,
+        "InactiveStateCode": 1,
         "TimeOutInMins": 2,
         "RetryCount": 3,
-        "SleepDuration": 2
+        "SleepDuration": 2,
+        "SubtractMinutesFromCurrentTime": -2
```
Comment (usha13529) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/appsettings.json:61]: Remove unused variable SubtractMinutesFromCurrentTime

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Services/HandleDeadLetterService.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+            await _callbackService.DeadLetterCallbackToD365UsingDataverse(entityPath, externalNotificationEntityWithStateCode, subscriptionRequestMessage);
+
+            _logger.LogInformation(EventIds.ENSSubscriptionMarkedAsInactiveCompleted.ToEventId(),
+                 "Process to mark subscription as inactive completed for SubscriptionId:{SubscriptionId}, _D365-Correlation-ID:{correlationId} and _X-Correlation-ID:{CorrelationId}", subscriptionId, subscriptionRequestMessage.D365CorrelationId, subscriptionRequestMessage.CorrelationId);
+
+            _logger.LogInformation(EventIds.ENSMoveDeadLetterContainerBlobStarted.ToEventId(),
+                 "Process to move dead letter container blob to destination container blob started for SubscriptionId:{SubscriptionId}, _D365-Correlation-ID:{correlationId} and _X-Correlation-ID:{CorrelationId}", subscriptionId, subscriptionRequestMessage.D365CorrelationId, subscriptionRequestMessage.CorrelationId);
```
Comment (usha13529) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Services/HandleDeadLetterService.cs:57]:
```suggestion
                 "Process to copy dead letter container blob to destination container blob started for SubscriptionId:{SubscriptionId}, _D365-Correlation-ID:{correlationId} and _X-Correlation-ID:{CorrelationId}", subscriptionId, subscriptionRequestMessage.D365CorrelationId, subscriptionRequestMessage.CorrelationId);
```

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Services/HandleDeadLetterService.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+            await _azureMessageQueueHelper.DeadLetterMoveBlob(_ensStorageConfiguration.Value, filePath);
+
+            _logger.LogInformation(EventIds.ENSMoveDeadLetterContainerBlobCompleted.ToEventId(),
+                 "Process to move dead letter container blob to destination container blob completed for SubscriptionId:{SubscriptionId}, _D365-Correlation-ID:{correlationId} and _X-Correlation-ID:{CorrelationId}", subscriptionId, subscriptionRequestMessage.D365CorrelationId, subscriptionRequestMessage.CorrelationId);
```
Comment (usha13529) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Services/HandleDeadLetterService.cs:62]:
```suggestion
                 "Process to copy dead letter container blob to destination container blob completed for SubscriptionId:{SubscriptionId}, _D365-Correlation-ID:{correlationId} and _X-Correlation-ID:{CorrelationId}", subscriptionId, subscriptionRequestMessage.D365CorrelationId, subscriptionRequestMessage.CorrelationId);
```

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/SubscriptionServiceJob.cs [hunk 4]
```diff
@@ -116,5 +121,35 @@ public async Task ProcessQueueMessage([QueueTrigger("%SubscriptionStorageConfigu
             _logger.LogInformation(EventIds.CreateSubscriptionRequestCompleted.ToEventId(),
                     "Subscription provisioning request Completed for SubscriptionId:{SubscriptionId} and _D365-Correlation-ID:{correlationId} and _X-Correlation-ID:{CorrelationId}", subscriptionMessage.SubscriptionId, subscriptionMessage.D365CorrelationId, subscriptionMessage.CorrelationId);
         }
+
+        public async Task ProcessBlobTrigger([BlobTrigger("%SubscriptionStorageConfiguration:StorageContainerName%/{filePath}")] Stream myBlob, string filePath)
+        {
+            string subscriptionId = GetSubscriptionIdName(filePath);
```
Comment (usha13529) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/SubscriptionServiceJob.cs:127]: Please rename method as below:
```suggestion
            string subscriptionId = GetSubscriptionId(filePath);
```

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/SubscriptionServiceJob.cs [hunk 5]
```diff
@@ -116,5 +121,35 @@ public async Task ProcessQueueMessage([QueueTrigger("%SubscriptionStorageConfigu
...
+            _logger.LogInformation(EventIds.ENSDeadLetterContainerJobRequestStart.ToEventId(),
+                    "External notification service - dead letter container webjob request started for SubscriptionId:{SubscriptionId}, FileName:{fileName}, _D365-Correlation-ID:{correlationId} and _X-Correlation-ID:{CorrelationId}", subscriptionId, fileName, subscriptionRequestMessage.D365CorrelationId, subscriptionRequestMessage.CorrelationId);
+
+            await _handleDeadLetterService.ProcessDeadLetter(filePath, subscriptionId, subscriptionRequestMessage);
+
+            _logger.LogInformation(EventIds.ENSDeadLetterContainerJobRequestCompleted.ToEventId(),
+                    "External notification service - dead letter container webjob request completed for SubscriptionId:{SubscriptionId}, FileName:{fileName}, _D365-Correlation-ID:{correlationId} and _X-Correlation-ID:{CorrelationId}", subscriptionId, fileName, subscriptionRequestMessage.D365CorrelationId, subscriptionRequestMessage.CorrelationId);
+        }
+
+        private static string GetSubscriptionIdName(string filePath)
```
Comment (usha13529) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/SubscriptionServiceJob.cs:141]:
```suggestion
        private static string GetSubscriptionId(string filePath)
```

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/SubscriptionServiceJob.cs [hunk 6]
```diff
@@ -116,5 +121,35 @@ public async Task ProcessQueueMessage([QueueTrigger("%SubscriptionStorageConfigu
             _logger.LogInformation(EventIds.CreateSubscriptionRequestCompleted.ToEventId(),
                     "Subscription provisioning request Completed for SubscriptionId:{SubscriptionId} and _D365-Correlation-ID:{correlationId} and _X-Correlation-ID:{CorrelationId}", subscriptionMessage.SubscriptionId, subscriptionMessage.D365CorrelationId, subscriptionMessage.CorrelationId);
         }
+
+        public async Task ProcessBlobTrigger([BlobTrigger("%SubscriptionStorageConfiguration:StorageContainerName%/{filePath}")] Stream myBlob, string filePath)
```
Comment (usha13529) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/SubscriptionServiceJob.cs:125]: Can we rename method as below please.
Also, can we remove unused parameter Stream myBlob?
```suggestion
        public async Task ProcessDeadLetterMessage([BlobTrigger("%SubscriptionStorageConfiguration:StorageContainerName%/{filePath}")] Stream myBlob, string filePath)
```

**Review by usha13529 — CHANGES_REQUESTED** [review]
Few comments to address please.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Helpers/IAzureMessageQueueHelper.cs [hunk 7]
```diff
@@ -9,5 +9,6 @@ public interface IAzureMessageQueueHelper
     {
         Task AddQueueMessage(SubscriptionStorageConfiguration ensStorageConfiguration, SubscriptionRequestMessage subscriptionRequestMessage);
         Task<HealthCheckResult> CheckMessageQueueHealth(string storageAccountConnectionString, string queueName);
+        Task DeadLetterMoveBlob(SubscriptionStorageConfiguration ensStorageConfiguration, string path);
```
[the remark this answers is not in the document]
Comment (ganpat-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Helpers/IAzureMessageQueueHelper.cs:12]: Thanks for highlighting, made required changes.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Helpers/AzureMessageQueueHelper.cs [hunk 8]
```diff
@@ -53,5 +54,30 @@ public async Task<HealthCheckResult> CheckMessageQueueHealth(string storageAccou
             else
                 return HealthCheckResult.Unhealthy("Azure message queue is unhealthy", new Exception($"Azure message queue {queueName} does not exists"));
         }
+
+        public async Task DeadLetterMoveBlob(SubscriptionStorageConfiguration ensStorageConfiguration, string path)
```
[the remark this answers is not in the document]
Comment (ganpat-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Helpers/AzureMessageQueueHelper.cs:58]: Done required changes.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService.UnitTests/Services/HandleDeadLetterServiceTest.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    [TestFixture]
+    public class HandleDeadLetterServiceTest
+    {
+        private ICallbackService _fakeCallbackService;
+        private IAzureMessageQueueHelper _fakeAzureMessageQueueHelper;
+        private IOptions<SubscriptionStorageConfiguration> _fakeEnsStorageConfiguration;
+        private ILogger<HandleDeadLetterService> _fakeLogger;
+        private IOptions<D365CallbackConfiguration> _fakeD365CallbackConfiguration;
+        private HandleDeadLetterService _handleDeadLetterService;
+        public string FilePath = "test";
```
[the remark this answers is not in the document]
Comment (ganpat-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService.UnitTests/Services/HandleDeadLetterServiceTest.cs:25]: Done required changes.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Configuration/D365CallbackConfiguration.cs [hunk 10]
```diff
@@ -11,12 +11,16 @@ public class D365CallbackConfiguration
...
+
+        public int InactiveStateCode { get; set; }     
 
         public int TimeOutInMins { get; set; }
 
         public int RetryCount { get; set; }
 
         public int SleepDuration { get; set; }
+
+        public int SubtractMinutesFromCurrentTime { get; set; }
```
[the remark this answers is not in the document]
Comment (ganpat-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Configuration/D365CallbackConfiguration.cs:24]: Done required changes.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Services/CallbackService.cs [hunk 11]
```diff
@@ -44,5 +44,28 @@ public async Task<HttpResponseMessage> CallbackToD365UsingDataverse(string exter
...
+            string accessToken = await _authTokenProvider.GetADAccessToken(subscriptionMessage);
+           
+            if (accessToken != string.Empty)
+            {
+                HttpResponseMessage httpResponse = await _callbackClient.GetCallbackD365Client(externalEntityPath, accessToken, externalNotificationEntityStateCode, subscriptionMessage.CorrelationId, CancellationToken.None);
+
+                if (httpResponse.StatusCode != HttpStatusCode.NoContent)
+                {
+                    _logger.LogError(EventIds.ErrorInDeadLetterCallbackToD365HttpClient.ToEventId(),
+                    "Callback to D365 failed as Status is inactive with :{StatusCode} and RequestUri:{RequestUri} and SubscriptionId:{SubscriptionId} and _D365-Correlation-ID:{correlationId} and _X-Correlation-ID:{CorrelationId}", httpResponse.StatusCode, httpResponse.RequestMessage.RequestUri, subscriptionMessage.SubscriptionId, subscriptionMessage.D365CorrelationId, subscriptionMessage.CorrelationId);
```
[the remark this answers is not in the document]
Comment (ganpat-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Services/CallbackService.cs:59]: Done required changes.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Services/HandleDeadLetterService.cs [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+            _azureMessageQueueHelper = azureMessageQueueHelper;
+            _ensStorageConfiguration = ensStorageConfiguration;
+            _logger = logger;
+            _d365CallbackConfiguration = d365CallbackConfiguration;
+        }
+
+        public async Task ProcessDeadLetter(string filePath, string subscriptionId, SubscriptionRequestMessage subscriptionRequestMessage)
+        {
+            _logger.LogInformation(EventIds.ENSSubscriptionMarkedAsInactiveStart.ToEventId(),
+                      "Process to mark subscription as inactive started for SubscriptionId:{SubscriptionId}, _D365-Correlation-ID:{correlationId} and _X-Correlation-ID:{CorrelationId}", subscriptionId, subscriptionRequestMessage.D365CorrelationId, subscriptionRequestMessage.CorrelationId);
```
[the remark this answers is not in the document]
Comment (ganpat-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Services/HandleDeadLetterService.cs:38]: Done required changes.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/appsettings.json (hunk 1 above)
Comment (ganpat-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/appsettings.json:61]: Done required changes.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Services/HandleDeadLetterService.cs (hunk 2 above)
Comment (ganpat-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Services/HandleDeadLetterService.cs:57]: Done required changes.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Services/HandleDeadLetterService.cs (hunk 3 above)
Comment (ganpat-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Services/HandleDeadLetterService.cs:62]: Done required changes.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/SubscriptionServiceJob.cs (hunk 4 above)
Comment (ganpat-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/SubscriptionServiceJob.cs:127]: Done required changes.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/SubscriptionServiceJob.cs (hunk 5 above)
Comment (ganpat-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/SubscriptionServiceJob.cs:141]: Done required changes.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/SubscriptionServiceJob.cs (hunk 6 above)
Comment (ganpat-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/SubscriptionServiceJob.cs:125]: Done required changes.
I have tried, but when I remove the unused parameter Stream myBlob it gives me an error.

**Review by usha13529 — APPROVED** [review]
Thanks for the changes @ganpat-tech. LGTM.

**Review by barrie-cooper — APPROVED** [review]
LGTM

**Pull request merged**